File: Patience Sort/PatienceSort.c
#include <stdio.h>

void patienceSort(int arr[], int n) {

    for(int i = 1; i < n; i++) {

        int key = arr[i];
        int j = i - 1;

        while(j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];
            j--;
        }

        arr[j + 1] = key;
    }
}

int main() {

    int arr[] = {9, 4, 6, 2, 8, 1};
    int n = 6;

    patienceSort(arr, n);

    for(int i = 0; i < n; i++)
        printf("%d ", arr[i]);

    return 0;
}

File: Patience Sort/PatienceSort.py
import heapq

def patience_sort(arr):

    piles = []

    for x in arr:

        placed = False

        for pile in piles:
            if x <= pile[-1]:
                pile.append(x)
                placed = True
                break

        if not placed:
            piles.append([x])

    heap = []

    for i, pile in enumerate(piles):
        heapq.heappush(heap, (pile.pop(), i))

    result = []

    while heap:

        value, i = heapq.heappop(heap)
        result.append(value)

        if piles[i]:
            heapq.heappush(heap, (piles[i].pop(), i))

    return result

arr = [9, 4, 6, 2, 8, 1]

print(patience_sort(arr))
